core/src/apps/zcash/signer.py

```diff
@@ -0,0 +1,155 @@
+from . import sig_hasher
+
+from micropython import const
+
+from trezor import log
+from trezor.utils import ensure
+from trezor.wire import ProcessError, DataError
+
+from apps.bitcoin.sign_tx.bitcoinlike import Bitcoinlike
+from apps.bitcoin.sign_tx import approvers
+from apps.common.writers import (
+    write_compact_size,
+    write_uint32_le,
+)
+
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from typing import Sequence, Tuple
+    from apps.common.coininfo import CoinInfo
+    from apps.bitcoin.sign_tx.tx_info import OriginalTxInfo, TxInfo
+    from apps.bitcoin.writers import Writer
+    from apps.bitcoin.sign_tx.sig_hasher import SigHasher
+    from apps.bitcoin.sign_tx.approvers import Approver
+    from trezor.utils import HashWriter
+    from trezor.messages import (
+        PrevTx, SignTx, TxInput, TxOutput,
+        TxAckInput, TxInfo, OriginalTxInfo,
+    )
+
+OVERWINTERED = const(0x8000_0000)
+
+class Zcash(Bitcoinlike):
+    def __init__(
+        self,
+        tx: SignTx,
+        keychain: Keychain,
+        coin: CoinInfo,
+        approver: Approver,
+    ) -> None:
+        ensure(coin.overwintered)
+        if tx.version != 5:
+            # seemless version upgrade
+            log.warning(__name__, "Transaction format version {} is not supported.".format(tx.version))
+            log.warning(__name__, "Switching to version 5.".format(tx.version))
+            tx.version = 5
+
+        super().__init__(tx, keychain, coin, approver)
+
+        self.tx_info.sig_hasher.initialize(self.tx_info.tx)
+
+    def create_sig_hasher(self) -> SigHasher:
+        return ZcashExtendedSigHasher()
+
+    def create_hash_writer(self) -> HashWriter:
+        # Replacement transactions are not supported
+        # so this should never be called.
+        raise NotImplementedError
+
+    async def step3_verify_inputs(self):
+        # Replacement transactions are not supported.
+
+        # We don't check prevouts, because BIP-341 techniques
+        # were adapted in ZIP-244 sighash algortihm.
+        # see: https://github.com/zcash/zips/issues/574
+        self.taproot_only = True  # turn on taproot behavior
+        await super().step3_verify_inputs()
+        self.taproot_only = False  # turn off taproot behavior
+
+    async def step5_serialize_outputs(self):
+        await super().step5_serialize_outputs()
+        await self.serialize_sapling_bundle()
+        await self.serialize_orchard_bundle()
```

Comment:
Semantically I think this actually belongs somewhere else. It's much more than just the outputs, it's both spends and outputs or entire actions. I'd suggest putting it in `write_tx_footer()` or `step7_finish()` or maybe even `step6_sign_segwit_inputs()`?